datafusion/src/catalog/catalog.rs

```diff
@@ -110,12 +111,23 @@ pub trait CatalogProvider: Sync + Send {
     fn schema(&self, name: &str) -> Option<Arc<dyn SchemaProvider>>;
 
     /// Adds a new schema to this catalog.
-    /// If a schema of the same name existed before, it is replaced in the catalog and returned.
+    ///
+    /// If a schema of the same name existed before, it is replaced in
+    /// the catalog and returned.
+    ///
+    /// By default returns a "Not Implemented" error
     fn register_schema(
         &self,
         name: &str,
         schema: Arc<dyn SchemaProvider>,
-    ) -> Option<Arc<dyn SchemaProvider>>;
+    ) -> Result<Option<Arc<dyn SchemaProvider>>> {
```

Comment:
this is the change -- return `Result` here (and add a default impl that returns error)